lab_2_tokenize_by_bpe/main.py

```diff
@@ -26,42 +40,124 @@ def collect_frequencies(
     :return: dictionary in the form of <token pair: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict):
+        return None
+
+    pair_dict = {}
+
+    for word in word_frequencies:
+        for index in range(len(word) - 1):
+            pair = (word[index], word[index + 1])
+            if pair not in pair_dict:
+                pair_dict[pair] = 0
+            pair_dict[pair] += word_frequencies[word]
+
+    return pair_dict
 
 
 def merge_tokens(
-    word_frequencies: dict[tuple[str, ...], int], pair: tuple[str, str]
+        word_frequencies: dict[tuple[str, ...], int], pair: tuple[str, str]
 ) -> dict[tuple[str, ...], int] | None:
     """
     Updates word frequency dictionary by replacing a pair of token with a merged one
     :param word_frequencies: dictionary in the form of <preprocessed word: number of occurrences>
     :param pair: a pair of tokens to be merged
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict) or not isinstance(pair, tuple):
+        return None
+
+    word_freq_updated = word_frequencies.copy()
+
+    for word in word_frequencies:
+        new_word = []
+        if pair[0] in word and pair[1] in word:
+            for index, token in enumerate(word):
+                if token == pair[1] and word[index - 1] == pair[0]:
+                    pass
+                elif token != pair[0]:
+                    new_word.append(token)
+                elif token == pair[0] and word[index + 1] != pair[1]:
+                    new_word.append(token)
+                elif token == pair[0] and word[index + 1] == pair[1]:
+                    new_word.append(pair[0] + pair[1])
+
+            value = word_freq_updated.pop(word)
+            word_freq_updated[tuple(new_word)] = value
+
+    return word_freq_updated
 
 
 def train(
-    word_frequencies: dict[tuple[str, ...], int] | None, num_merges: int
+        word_frequencies: dict[tuple[str, ...], int] | None, num_merges: int
 ) -> dict[tuple[str, ...], int] | None:
     """
     Creates required number of new tokens by merging existing ones
     :param word_frequencies: dictionary of a kind <preprocessed word: number of occurrences>
     :param num_merges: required number of new tokens
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict) or not isinstance(num_merges, int):
+        return None
+
+    pair_dict = count_tokens_pairs(word_frequencies)
+    if pair_dict is None:
+        return None
+
+    for _ in range(num_merges):
+        max_value = max(pair_dict.values())
+        pair_list = [key for key, value in pair_dict.items() if value == max_value]
+
+        len_list = []
+        for pair in pair_list:
+            pair_len = len(''.join(list(pair)))
+            len_list.append(pair_len)
+        max_len = max(len_list)
```

Comment:
Consider simply using function `max` with the argument `key`

Reply:
You can even specify the alphabetic order of choosing between several items with the same length

Reply:
I don't understand how I do this because I have to choose a few pairs with largest length and then choose the one that is smaller lexicographically. But how do I have a few pairs with largest length when using function max with the argument key? (Doesn't it return only one value? I can create a cycle but I how do I know how many pairs with the largest length do I have?) Or how do I implement both keys to sort length and alphabetic order at the same time?

Reply:
nvm, I got it